README.md
# check-link

## API

- scanSite
- scanUrls
- checkUrl
- getLinks

## Dependencies

npx download-size -f package.json

```bash
(check-link@1.2.0):
  dependencies:                                                                                                                                                     
   array-random-item@1.0.2: 1.70 KiB                                                                                                                                
   await-timeout@1.1.1: 4.14 KiB                                                                                                                                    
   axios@0.27.2: 182.51 KiB                                                                                                                                         
   cheerio@1.0.0-rc.12: 380.29 KiB                                                                                                                                  
   keyv@4.5.2: 10.98 KiB                                                                                                                                            
   random@3.0.6: 113.42 KiB                                                                                                                                            top-user-agents@1.0.52: 640.38 KiB
  devDependencies:                                                                                                                                                     jest@27.5.1: 5.66 MiB
All dependencies: 6.96 MiB
```


CHANGELOG.md
# 📋 Changelog

## 1.2.0
- Add `redirectsCount` and `redirectedUrl` in case of redirects for `checkUrl()`
- Update dependencies

## 1.1.1
- Update dependencies due to security reasons

## 1.1.0
- Add optional param maxMinutes to scanSite()

## 1.0.0
- Initial release with:
  - scanSite
  - scanUrls
  - checkUrl
  - getLinks

src/checkUrl.spec.js
const checkUrl = require('./checkUrl')

describe('checkUrl()', () => {
  jest.setTimeout(12000)

  it('is defined', () => {
    expect(checkUrl).toBeDefined();
    expect(typeof checkUrl).toEqual('function');
  })

  it('check a valid url', async () => {
    const URL = 'https://davidburgos.blog/';
    let out;
    let error;
    try {
      out = await checkUrl(URL);
    } catch (e) {
      error = e;
    } finally {
      expect(error).toBeUndefined();
      expect(out).toBeDefined();
      expect(out.url).toEqual(URL);
      expect(out.status).toEqual(200);
      expect(out.isOK).toEqual(true);
      expect(out.responseTime).toBeGreaterThan(0);
    }
  });

	it('check a redirect url', async () => {
		const URL = 'http://davidburgos.blog/';
		let out;
		let error;
		try {
			out = await checkUrl(URL);
		} catch (e) {
			error = e;
		} finally {
			expect(error).toBeUndefined();
			expect(out).toBeDefined();
			expect(out.url).toEqual(URL);
			expect(out.status).toEqual(200);
			expect(out.isOK).toEqual(true);
			expect(out.responseTime).toBeGreaterThan(0);
			expect(out.redirectsCount).toBeGreaterThan(0);
			expect(out.redirectedUrl).toEqual(URL.replace('http','https'));
		}
	});

  it('check a not found url', async () => {
    const URL = 'https://davidburgos.blog/not-found';
    let out;
    let error;
    try {
      out = await checkUrl(URL);
    } catch (e) {
      error = e;
    } finally {
      expect(error).toBeUndefined();
      expect(out).toBeDefined();
      expect(out.url).toEqual(URL);
      expect(out.status).toEqual(404);
      expect(out.isOK).toEqual(false);
      expect(out.responseTime).toBeGreaterThan(0);
    }
  });

  it('check a not found url', async () => {
    const URL = 'https://glasslytics.com/';
    let out;
    let error;
    try {
      out = await checkUrl(URL);
    } catch (e) {
      error = e;
    } finally {
      expect(error).toBeUndefined();
      expect(out).toBeDefined();
      expect(out.url).toEqual(URL);
      expect(out.status).toEqual(500);
      expect(out.isOK).toEqual(false);
      expect(out.responseTime).toBeGreaterThan(0);
    }
  });

  it('deal with a timeout as valid', async () => {
    const URL = 'https://www.seattletimes.com/business/amazon/amazon-posts-record-sales-and-profit-with-no-slowing-expected/';
    let out;
    let error;
    try {
      out = await checkUrl(URL);
    } catch (e) {
      error = e;
    } finally {
      expect(error).toBeUndefined();
      expect(out).toBeDefined();
      expect(out.url).toEqual(URL);
      expect(out.status).toEqual(500);
      expect(out.isOK).toEqual(true);
      expect(out.responseTime).toBeGreaterThan(0);
    }
  });
});

src/getLinks.spec.js
const getLinks = require('./getLinks')

describe('getLinks()', () => {
  it('is defined', () => {
    expect(getLinks).toBeDefined();
    expect(typeof getLinks).toEqual('function');
  })

  it('get links array from a url', async () => {
    const URL = 'https://davidburgos.blog/';
    let out;
    let error;
    try {
      out = await getLinks(URL);
    } catch (e) {
      error = e;
    } finally {
      expect(error).toBeUndefined();
      expect(out).toBeDefined();
      expect(out.url).toEqual(URL);
      expect(out.links.length).toBeGreaterThanOrEqual(0);
      expect(out.countLinks).toBeGreaterThan(0);
      expect(out.responseTime).toBeGreaterThan(0);
      expect(out.executionTime).toBeGreaterThan(0);
    }
  });
});

src/scanSite.spec.js
const scanSite = require('./scanSite')

describe('scanSite()', () => {
	jest.setTimeout(10000);
	it('is defined', () => {
		expect(scanSite).toBeDefined();
		expect(typeof scanSite).toEqual('function');
	})

	it('get all links from a url in recursive mode', async () => {
		const URL = 'https://davidburgos.blog/';
		const MAX_PAGES = 2;
		let error;
		let pagesDone = 0;
		try {
			const options = {
				recursive: true,
				maxPages: MAX_PAGES,
				onPageError: (e) => {
					error = e;
				},
				onPageDone: (result) => {
					expect(result.links.length).toBeGreaterThanOrEqual(0);
					expect(result.countLinks).toBeGreaterThan(0);
					expect(result.responseTime).toBeGreaterThan(0);
					expect(result.executionTime).toBeGreaterThan(0);
					expect(pagesDone).toBeGreaterThanOrEqual(0);
					expect(pagesDone).toBeLessThanOrEqual(MAX_PAGES);
					pagesDone += 1;
				},
				onEnd: (result) => {
					expect(result.pages).toBeGreaterThan(0);
					expect(result.executionTime).toBeGreaterThan(0);
				}
			}
			await scanSite(URL, options);
		} catch (e) {
			error = e;
		} finally {
			expect(error).toBeUndefined();
		}
	});

	it('get the error from a link', async () => {
		const URL = 'https://davidburgos.blog/tag/not-found';
		let error;
		try {
			const options = {
				recursive: false,
				onPageError: (e) => {
					expect(e).toBeDefined()
					expect(e.url).toEqual(URL)
					expect(e.error).toBeDefined()
					expect(e.error.response).toBeDefined()
					expect(e.error.response.status).toEqual(404)
				}
			}
			await scanSite(URL, options);
		} catch (e) {
			console.log('>> catch', e);
			error = e;
		} finally {
			expect(error).toBeUndefined();
		}
	})

	it('scan site until maxMinutes param', async () => {
		const URL = 'https://davidburgos.blog/';
		const MAX_SECONDS = 5;
		let error;
		try {
			const options = {
				recursive: true,
				maxPages: 2,
				maxMinutes: MAX_SECONDS/60,
				onEnd: (result) => {
					expect(result.pages).toBeGreaterThan(0);
					expect(result.executionTime).toBeGreaterThan(MAX_SECONDS);
					expect(result.executionTime).toBeLessThan(MAX_SECONDS*2);
				}
			}
			await scanSite(URL, options);
		} catch (e) {
			error = e;
		} finally {
			expect(error).toBeUndefined();
		}
	})
});

src/checkUrl.js
const axios = require('axios');
const userAgents = require('top-user-agents');
const randomItem = require('array-random-item');

module.exports = async function (url, options = {}) {
  const requestConfig = {
    headers: { 'User-Agent': randomItem(userAgents) },
    timeout: 5 * 1000
  };
  let start;
  let responseTime;
  let isOK;
  let status;
	let redirectsCount = 0;
	let redirectedUrl;
  try {
    start = new Date().getTime();
    const out = await axios.head(url, requestConfig);
		redirectsCount = out?.request?._redirectable?._redirectCount;
		redirectedUrl = out?.request?._redirectable?._currentUrl;
    responseTime = (new Date().getTime() - start) / 1000;
    isOK = out && (out.status === 200 || out.statusText === 'OK');
    status = out.status;
  } catch (e) {
    responseTime = (new Date().getTime() - start) / 1000;
    isOK = false;
    status = e && e.response && e.response.status || 500;
  }
  if (!isOK) {
    try {
      start = new Date().getTime();
      const out = await axios.get(url, requestConfig);
	    redirectsCount = out?.request?._redirectable?._redirectCount;
	    redirectedUrl = out?.request?._redirectable?._currentUrl;
      responseTime = (new Date().getTime() - start) / 1000;
    } catch (e) {
      responseTime = (new Date().getTime() - start) / 1000;
      isOK = e.code === 'ECONNABORTED' || false;
      status = e && e.response && e.response.status || 500;
    }
  }

  const output = {
    url,
    status,
    isOK,
    responseTime
  };
	if (redirectsCount > 0) {
		output.redirectsCount = redirectsCount;
		output.redirectedUrl = redirectedUrl;
	}
  return output;
}

src/getLinks.js
const axios = require('axios');
const cheerio = require('cheerio');
const userAgents = require('top-user-agents');
const randomItem = require('array-random-item');

module.exports = async function (url, options = {}) {
  const {
    unique = true
  } = options;
  const requestConfig = {
    headers: { 'User-Agent': randomItem(userAgents) }
  };
  const start = new Date().getTime();
  const html = await axios.get(url, requestConfig).then((res) => res.data);
  const responseTime = (new Date().getTime()-start)/1000;
  const $ = cheerio.load(html);
  const anchors = $('a[href]');
  const links = [];
  const countLinks = anchors.length;
  $(anchors).each(function (i, link) {
    const href = $(link).attr('href');
    const fullPath = new URL(href, url).href;
    const isDifferent = url !== fullPath;
    const addLink = isDifferent && (!unique || (unique && !links.includes(fullPath)));
    if (addLink) {
      links.push(fullPath);
    }
  });
  const output = {
    url,
    links,
    countLinks,
    responseTime,
    executionTime: (new Date().getTime()-start)/1000
  };
  return output;
}

src/scanSite.js
const Timeout = require('await-timeout');
const random = require('random');
const getLinks = require('./getLinks');

module.exports = async function (url, options = {}) {
    const {
        recursive = false,
        minInterval = 1000,
        maxInterval = 3000,
        maxPages = false,
	      maxMinutes = false,
        onPageDone,
        onPageError,
        onEnd
    } = options;
    const startTime = new Date().getTime();
    let pages = 0;
    let lastRequestTime = new Date().getTime();
    let result;
    try {
        result = await getLinks(url, options);
    } catch (e) {
        if (onPageError && typeof onPageError === 'function') {
            onPageError({ url, error: e });
        }
    }
    if (!recursive) {
        if (onPageDone && typeof onPageDone === 'function') {
            onPageDone(result);
        }
        if (onEnd && typeof onEnd === 'function') {
            onEnd({ pages, executionTime: (new Date().getTime() - startTime) / 1000 });
        }
    }
    const nextPages = [];
    result && result.links.forEach((link) => {
        const isSameDomain = link.startsWith(url);
        if (isSameDomain) {
            nextPages.push({ page: url, link });
        }
    });
    for (const page of nextPages) {
        try {
            const now = new Date().getTime();
            const diffTime = now - lastRequestTime;
            if (diffTime < minInterval) {
                const maxTimeLeft = maxInterval - diffTime;
                const minTimeLeft = minInterval - diffTime;
                const timeLeft = random.int(minTimeLeft, maxTimeLeft);
                await Timeout.set(timeLeft);
            }
            lastRequestTime = new Date().getTime();
            const lastOutput = await getLinks(page.link, options);
            pages += 1;
            if (onPageDone && typeof onPageDone === 'function') {
                onPageDone(lastOutput);
            }
						const executionTime = (new Date().getTime() - startTime) / 1000;
						const doneDueMaxPages = maxPages && pages >= maxPages;
						const doneDueTime = maxMinutes && executionTime/60 >= maxMinutes;
						const mustEndNow = doneDueMaxPages || doneDueTime;
            if (mustEndNow) {
                if (onEnd && typeof onEnd === 'function') {
                    return onEnd({ pages, executionTime });
                }
            }
            lastOutput.links.forEach((link) => {
                const isSameDomain = link.startsWith(url);
                const alreadyIncluded = nextPages.includes(link);
                const proceed = isSameDomain && !alreadyIncluded;
                if (proceed) {
                    nextPages.push({ page: page.link, link });
                }
            });
        } catch (e) {
            if (onPageError && typeof onPageError === 'function') {
                onPageError({ page: page.page, url: page.link, error: e });
            }
        }
    }
    const output = { pages, executionTime: (new Date().getTime() - startTime) / 1000 };
    if (onEnd && typeof onEnd === 'function') {
        onEnd(output);
    }
    return output;
}

src/scanUrls.js
const Timeout = require('await-timeout');
const random = require('random');
const checkUrl = require('./checkUrl');
const Keyv = require('keyv');
const keyv = new Keyv();
const ONE_HOUR = 1 * 60 * 60 * 1000;
const HOST_CACHE_EXPIRATION = ONE_HOUR;

module.exports = async function (urls = [], options = {}) {
  const {
    minInterval = 1000,
    maxInterval = 3000,
    onUrlError,
    onUrlDone,
    onEnd
  } = options;
  const queue = {};
  const startTime = new Date().getTime();
  let urlsDone = 0;
  urls.forEach((url) => {
    let hostname;
    try {
      hostname = new URL(url).hostname;
      if (queue[hostname]) {
        queue[hostname].push(url);
      } else {
        queue[hostname] = [url];
      }
    } catch (e) {
      console.log('> newUrl catch', url, e);
    }
  });
  const promises = Object.keys(queue).map(async (host) => {
    for (const url of queue[host]) {
      let lastHostRequestAt;
      try {
        lastHostRequestAt = await keyv.get(host);
      } catch (e) {
        console.log('> newUrl catch', url, e);
      }
      try {
        if (lastHostRequestAt) {
          const now = new Date().getTime();
          const diffTime = now - lastHostRequestAt;
          if (diffTime < minInterval) {
            const maxTimeLeft = maxInterval - diffTime;
            const minTimeLeft = minInterval - diffTime;
            const timeLeft = random.int(minTimeLeft, maxTimeLeft);
            await Timeout.set(timeLeft);
          }
        }

        lastRequestTime = new Date().getTime();
        await keyv.set(host, lastRequestTime, HOST_CACHE_EXPIRATION);
        const lastOutput = await checkUrl(url, options);
        urlsDone += 1;
        if (onUrlDone && typeof onUrlDone === 'function') {
          onUrlDone(lastOutput);
        }
      } catch (e) {
        if (onUrlError && typeof onUrlError === 'function') {
          onUrlError({ url, error: e });
        }
      }
    }
  });
  await Promise.all(promises);
  const output = { urlsDone, executionTime: (new Date().getTime() - startTime) / 1000 };
  if (onEnd && typeof onEnd === 'function') {
    onEnd(output);
  }
  return output;
}